File: functions/src/challenge_board_presets.test.ts
/**
 * Parity vectors for Challenge board presets — keep in sync with
 * test/challenge/challenge_board_presets_test.dart
 */
import {
  CHALLENGE_BOARD_PRESETS,
  DEFAULT_CHALLENGE_PRESET_ID,
  initialGameStateForPreset,
  InvalidChallengePresetError,
  resolveChallengePreset,
} from './challenge_board_presets';
import { GameRules } from './game_rules';

interface PresetVector {
  id: string;
  rows: number;
  cols: number;
  totalBoxes: number;
  initialLegalMoves: number;
  disabledCells: string[];
}

const PRESET_VECTORS: PresetVector[] = [
  {
    id: 'challenge_classic',
    rows: 6,
    cols: 6,
    totalBoxes: 25,
    initialLegalMoves: 60,
    disabledCells: [],
  },
  {
    id: 'challenge_blitz',
    rows: 4,
    cols: 4,
    totalBoxes: 9,
    initialLegalMoves: 24,
    disabledCells: [],
  },
  {
    id: 'challenge_fortress',
    rows: 5,
    cols: 5,
    totalBoxes: 7,
    initialLegalMoves: 22,
    disabledCells: [
      '1_1', '1_2', '1_3',
      '2_1', '2_2', '2_3',
      '3_1', '3_2', '3_3',
    ],
  },
];

function assert(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

function runTests(): void {
  assert(
    CHALLENGE_BOARD_PRESETS.length === 3,
    'Expected exactly 3 challenge presets',
  );

  for (const vector of PRESET_VECTORS) {
    const preset = resolveChallengePreset(vector.id);
    assert(preset.rows === vector.rows, `${vector.id} rows mismatch`);
    assert(preset.cols === vector.cols, `${vector.id} cols mismatch`);
    assert(
      [...preset.disabledCells].sort().join(',') === vector.disabledCells.sort().join(','),
      `${vector.id} disabledCells mismatch`,
    );

    const state = initialGameStateForPreset(preset);
    assert(state.totalBoxes === vector.totalBoxes, `${vector.id} totalBoxes mismatch`);
    assert(
      GameRules.legalMoves(state).length === vector.initialLegalMoves,
      `${vector.id} initialLegalMoves mismatch`,
    );
    assert(state.currentPlayerId === 'A', `${vector.id} should start on player A`);
    assert(!state.isOver, `${vector.id} should not start over`);
  }

  assert(
    resolveChallengePreset(undefined).id === DEFAULT_CHALLENGE_PRESET_ID,
    'Missing preset id should default to classic',
  );
  assert(
    resolveChallengePreset('').id === DEFAULT_CHALLENGE_PRESET_ID,
    'Empty preset id should default to classic',
  );

  let rejected = false;
  try {
    resolveChallengePreset('invalid');
  } catch (err) {
    rejected = err instanceof InvalidChallengePresetError;
  }
  assert(rejected, 'Invalid preset id should throw InvalidChallengePresetError');

  console.log('challenge_board_presets.test: all assertions passed');
}

runTests();

File: test/challenge/challenge_board_presets_test.dart
import 'package:dot_clash/features/game/domain/models/ai_preset.dart';
import 'package:dot_clash/features/game/domain/models/game_state.dart';
import 'package:dot_clash/features/game/domain/rules/game_rules.dart';
import 'package:flutter_test/flutter_test.dart';

/// Parity vectors — keep in sync with functions/src/challenge_board_presets.test.ts
void main() {
  final fortressDisabled = AiPreset.byId('fortress')!.disabledCells;

  group('Challenge board preset parity', () {
    test('challenge_classic initial state matches server vectors', () {
      final state = GameState.initial(rows: 6, cols: 6, playerIds: const ['A', 'B']);
      expect(state.totalBoxes, 25);
      expect(GameRules.legalMoves(state).length, 60);
      expect(state.currentPlayerId, 'A');
      expect(state.isOver, isFalse);
    });

    test('challenge_blitz initial state matches server vectors', () {
      final state = GameState.initial(rows: 4, cols: 4, playerIds: const ['A', 'B']);
      expect(state.totalBoxes, 9);
      expect(GameRules.legalMoves(state).length, 24);
      expect(state.currentPlayerId, 'A');
      expect(state.isOver, isFalse);
    });

    test('challenge_fortress initial state matches server vectors', () {
      final state = GameState.initial(
        rows: 5,
        cols: 5,
        playerIds: const ['A', 'B'],
        disabledCells: fortressDisabled.toSet(),
      );
      expect(state.totalBoxes, 7);
      expect(GameRules.legalMoves(state).length, 22);
      expect(state.currentPlayerId, 'A');
      expect(state.isOver, isFalse);
      expect(
        state.disabledCells,
        {
          '1_1', '1_2', '1_3',
          '2_1', '2_2', '2_3',
          '3_1', '3_2', '3_3',
        },
      );
    });
  });
}
